logical/plugin/backend_client.go

```diff
@@ -0,0 +1,226 @@
+package plugin
+
+import (
+	"net/rpc"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vault/logical"
+	log "github.com/mgutz/logxi/v1"
+)
+
+// backendPluginClient implements logical.Backend and is the
+// go-plugin client.
+type backendPluginClient struct {
+	broker       *plugin.MuxBroker
+	client       *rpc.Client
+	pluginClient *plugin.Client
+
+	system logical.SystemView
+	logger log.Logger
+}
+
+// HandleRequestArgs is the args for HandleRequest method.
+type HandleRequestArgs struct {
+	StorageID uint32
+	Request   *logical.Request
+}
+
+// HandleRequestReply is the reply for HandleRequest method.
+type HandleRequestReply struct {
+	Response *logical.Response
+	Error    *plugin.BasicError
+}
+
+// SpecialPathsReply is the reply for SpecialPaths method.
+type SpecialPathsReply struct {
+	Paths *logical.Paths
+}
+
+// SystemReply is the reply for System method.
+type SystemReply struct {
+	SystemView logical.SystemView
+	Error      *plugin.BasicError
+}
+
+// HandleExistenceCheckArgs is the args for HandleExistenceCheck method.
+type HandleExistenceCheckArgs struct {
+	StorageID uint32
+	Request   *logical.Request
+}
+
+// HandleExistenceCheckReply is the reply for HandleExistenceCheck method.
+type HandleExistenceCheckReply struct {
+	CheckFound bool
+	Exists     bool
+	Error      *plugin.BasicError
+}
+
+// ConfigureArgs is the args for Configure method.
+type ConfigureArgs struct {
+	StorageID uint32
+	LoggerID  uint32
+	SysViewID uint32
+	Config    map[string]string
+}
+
+// ConfigureReply is the reply for Configure method.
+type ConfigureReply struct {
+	Error *plugin.BasicError
+}
+
+// TypeReply is the reply for the Type method.
+type TypeReply struct {
+	Type logical.BackendType
+}
+
+func (b *backendPluginClient) HandleRequest(req *logical.Request) (*logical.Response, error) {
+	// Shim logical.Storage
+	id := b.broker.NextId()
+	go b.broker.AcceptAndServe(id, &StorageServer{
```

Comment:
Using `req.Storage` is a decision from the very beginnings of Vault so I only have some guesses as to why it's that way, but I think it's to provide flexibility in case the storage should be different per request. For instance, it's not actually how `cubbyhole` is written but you can imagine that the cubbyhole backend would just have different storage views (via say a barrier subview) per token to ensure prefixing is done properly.

Realistically, though, that functionality has never been used. Probably you can just shove in the same underlying storage for all requests.